include/fastdds/rtps/security/logging/Logging.h

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Canonical ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*!
+ * @file Logging.h
+ */
+#ifndef _FASTDDS_RTPS_SECURITY_LOGGING_LOGGING_H_
+#define _FASTDDS_RTPS_SECURITY_LOGGING_LOGGING_H_
+
+#include "fastdds/rtps/security/logging/LogOptions.h"
+#include "fastdds/rtps/security/exceptions/SecurityException.h"
+
+namespace eprosima {
+namespace fastrtps {
+namespace rtps {
+namespace security {
+
+/**
+ * @brief The LoggerListener class
+ */
+class LoggerListener
+{
+  LoggerListener() = default;
+  ~LoggerListener() = default;
+};
+
+/**
+ * @brief Base class for all security logging plugins.
+ */
+class Logging
+{
+public:
+
+  Logging();
+  virtual ~Logging() = default;
+
+  /**
+   * @brief set_log_options
+   * @param log_options
+   * @return TRUE if successful
+   */
+  bool set_log_options(const LogOptions& log_options, SecurityException& exception);
+
+  /**
+   * @brief get_log_options
+   * @param log_options
+   * @param exception
+   * @return
+   */
+  bool get_log_options(LogOptions& log_options, SecurityException& exception) const;
+
+  /**
+   * @brief enable_logging
+   */
+  bool enable_logging(SecurityException& exception);
+
+  /**
+   * @brief set_listener
+   * @param listener
+   * @return
+   */
+  bool set_listener(LoggerListener* listener, SecurityException& exception);
+
+  /**
+   * @brief log
+   * @param log_level
+   * @param message
+   * @param category
+   */
+  void log(const EventLogLevel event_log_level,
+           const std::string& message,
+           const std::string& category,
+           SecurityException& exception);
+
+protected:
+
+  /**
+   * @brief log_impl
+   * @param message
+   * @param category
+   * @param exception
+   */
+  virtual void log_impl(const std::string& message,
+                        const std::string& category,
+                        SecurityException& exception);
+
+  LoggerListener* listener_;
+
+  bool logging_enabled_ = false;
+  bool options_set_ = false;
+
+  LogOptions log_options_;
```

Comment:
If we leave these here subclasses will be able to modify them. Do they need the ability to do that?

Reply:
For the `listener_` not sure, for the rest, likely not. However subclasses my need to read them for some reason, should I make them private and add some getter functions?

Reply:
Yeah I suggest making everything private until we learn that it _cannot_ be, at which time we promote to the most protected section that works. If we learn the access required is read-only, then we make a protected getter and leave the variable private. If we learn the access required is read-write, then we can tackle that on a case-by-case basis.

Reply:
I'd rather make them private and implement the getters from the get go. We may not be around when a subclass will be developed and it could be too tempting for someone to change that private to protected.

Reply:
Fair enough.